File: daily/2026/3756_Concatenate_Non_Zero_Digits_and_Multiply_by_Sum_II.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

const int mod = 1e9 + 7;
const int MAX_N = 100001;
long long pow10[MAX_N];

int init = [](){
    pow10[0] = 1;
    for (int i = 1; i < MAX_N; i++) pow10[i] = pow10[i-1] * 10 % mod;
    return 0;
}();

class Solution {
public:
    vector<int> sumAndMultiply(string s, vector<vector<int>>& queries) {
        int n = s.length();
        vector<long long> concatenated(n + 1);
        vector<int> acc(n + 1);
        vector<int> cnt(n + 1);
        for (int i = 0; i < n; i++){
            int d = s[i] - '0';
            if (d > 0) concatenated[i + 1] = (concatenated[i] * 10 + d) % mod;
            else concatenated[i + 1] = concatenated[i];
            acc[i + 1] = acc[i] + d;
            cnt[i + 1] = cnt[i] + (d > 0);
        }
        int q = queries.size();
        vector<int> ans(q);
        for (int i = 0; i < q; i++){
            int s = queries[i][0];
            int e = queries[i][1] + 1;
            int length = cnt[e] - cnt[s];
            long long x = (concatenated[e] - (concatenated[s] * pow10[length]) % mod + mod) % mod;
            ans[i] = (x * (acc[e] - acc[s])) % mod;
        }
        return ans;
    }
};

int main(){
    // string s = "10203004";
    // vector<vector<int>> queries = {{0,7},{1,3},{4,6}};
    // string s = "1000";
    // vector<vector<int>> queries = {{0,3},{1,1}};
    string s = "9876543210";
    vector<vector<int>> queries = {{0,9}};
    Solution S;
    vector<int> ans = S.sumAndMultiply(s, queries);
    for (int i : ans) cout << i << " ";
    cout << endl;
    return 0;
}

File: daily_python/3756_Concatenate_Non_Zero_Digits_and_Multiply_by_Sum_II.py
import os
import sys
from typing import List

mod = 10 ** 9 + 7
pow10 = [1] * 100001
for i in range(1, 100001):
    pow10[i] = pow10[i-1] * 10 % mod

class Solution:
    def sumAndMultiply(self, s: str, queries: List[List[int]]) -> List[int]:
        n = len(s)
        acc = [0] * (n + 1)
        x = [0] * (n + 1)
        cnt = [0] * (n + 1)
        for i, c in enumerate(s):
            d = int(c)
            acc[i + 1] = acc[i] + d
            x[i + 1] = x[i] if d == 0 else (x[i] * 10 + d) % mod
            cnt[i + 1] = cnt[i] + (d > 0)
        m = len(queries)
        ans = [0] * m
        for i in range(m):
            l = queries[i][0]
            r = queries[i][1] + 1
            ans[i] = (x[r] - x[l] * pow10[cnt[r] - cnt[l]]) * (acc[r] - acc[l]) % mod
        return ans

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            queries = [list(map(int, query.split(','))) for query in lines[1].split('=')[1].strip()[2:-3].split('},{')]
            dataset.append((s, queries))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s, queries in dataset:
        results.append(solution.sumAndMultiply(s, queries))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')